### PR TITLE
Manually implement `is_zipfile`

### LINKED ISSUE #25214: torch.load issue on loading file created by torch.save
## 🐛 Bug

For some particular data, `torch.load()` can not properly load a file dumped by `torch.save()`.

Root cause:
The file was mistakenly identified as a zip file by `zipfile.is_zipfile()` at line 557 of `torch\serialization.py`, due to the collision of magic number.

## To Reproduce

1. Use any dumped tensor from https://github.com/nozomifans/reccell/tree/master/issue
2. Load and save data in uint8, then try to load it back
```
import torch

data = torch.load('E15_s1_08.pt')

torch.save(data.type(torch.uint8),'E15_s1_08_test.pt') 

data = torch.load('E15_s1_08_test.pt')
```
3. Traceback
```
Traceback (most recent call last):

  File "<ipython-input-12-7f84586dd470>", line 1, in <module>
    runfile('/Documents/workspace/loadissue/loadtest.py', wdir='/Documents/workspace/loadissue')

  File "\AppData\Local\Continuum\anaconda3\envs\p36-dev\lib\site-packages\spyder_kernels\customize\spydercustomize.py", line 827, in runfile
    execfile(filename, namespace)

  File "\AppData\Local\Continuum\anaconda3\envs\p36-dev\lib\site-packages\spyder_kernels\customize\spydercustomize.py", line 110, in execfile
    exec(compile(f.read(), filename, 'exec'), namespace)

  File "/Documents/workspace/loadissue/loadtest.py", line 18, in <module>
    data = torch.load('E15_s1_08_test.pt')

  File "\AppData\Local\Continuum\anaconda3\envs\p36-dev\lib\site-packages\torch\serialization.py", line 386, in load
    return _load(f, map_location, pickle_module, **pickle_load_args)

  File "\AppData\Local\Continuum\anaconda3\envs\p36-dev\lib\site-packages\torch\serialization.py", line 559, in _load
    raise RuntimeError("{} is a zip archive (did you mean to use torch.jit.load()?)".format(f.name))

RuntimeError: E15_s1_08_test.pt is a zip archive (did you mean to use torch.jit.load()?)
```

## Expected behavior

torch.load should load the tensor dumped by torch.save

## Environment

 - PyTorch Version (e.g., 1.0): 1.2
 - OS (e.g., Linux): Windows 10 or CentOS 7.6.1810
 - How you installed PyTorch (`conda`, `pip`, source): Conda
 - Build command you used (if compiling from source):
 - Python version: 3.6 or 3.7
 - CUDA/cuDNN version:
 - GPU models and configuration: CPU only or 4x2080
 - Any other relevant information:

## Additional context

Current workaround is to save such collision into float32 to avoid the collision. Is that possible to provide an argument to flag this file is not a zip?

### PR DESCRIPTION
**Pull request by driazati**
The default implementation is lenient in that it recognizes a zipfile if the magic number appears anywhere in the archive. So if someone has the bytes `PK\x03\x04` in a tensor, it gets recognized as a zipfile. See https://bugs.python.org/issue28494
    
This implementation only checks the first 4 bytes of the file for the zip magic number. We could also copy https://github.com/python/cpython/pull/5053's fix, but that seems like overkill.
    
Fixes #25214

Differential Revision: [D17102516](https://our.internmc.facebook.com/intern/diff/17102516/)

### CODE REVIEW

**Review by zou3519 — CHANGES_REQUESTED** [review]
Needs test unless it's hard to repro the bug. However, it sounds like one just needs to serialize a tensor with the magic numbers?

**Review by ezyang — CHANGES_REQUESTED** [review]
As Richard said, needs test. Lint failures are real too. Also, shouldn't we be using `b''` strings?

**Comment by driazati** [comment]
`b''` doesn't do anything in Python 2, `ord()` takes care of turning things into bytes

File: torch/serialization.py [hunk 1]
```diff
@@ -44,6 +43,43 @@ def mkdtemp():
 _package_registry = []
 
 
+def _is_zipfile(f):
+    # This is a stricter implmentation than zipfile.is_zipfile().
```
Comment (ezyang) [inline torch/serialization.py]: nit: implementation

**Review by ezyang — APPROVED** [review]

**Review by zou3519 — APPROVED** [review]

**Pull request closed without merging**